doc/man/gnome-session-quit.1
.\"
.\" cinnamon-session-quit manual page.
.\" (C) 2000 Miguel de Icaza (miguel@helixcode.com)
.\" (C) 2009-2010 Vincent Untz (vuntz@gnome.org)
.\"
.TH GNOME-SESSION-QUIT 1 "GNOME"
.SH NAME
cinnamon-session-quit \- End the current GNOME session
.SH SYNOPSIS
.B cinnamon-session-quit [\-\-logout|\-\-power-off|\-\-reboot] [\-\-force] [\-\-no-prompt]
.SH DESCRIPTION
The \fIcinnamon-session-quit\fP program can be used to end a GNOME session.
.SH OPTIONS
The following options are supported:
.TP
.I "--logout"
Prompt the user to confirm logout. This is the default behavior.
.TP
.I "--power-off"
Prompt the user to confirm system power off.
.TP
.I "--reboot"
Prompt the user to confirm system reboot.
.TP
.I "--force"
Ignore any inhibitors.
.TP
.I "--no-prompt"
End the session without user interaction. This only works with \fI--logout\fP.
.SH SEE ALSO
.BR cinnamon-session(1)


NEWS
=============
Version 3.6.2
=============

 Session Manager

	* initialize systemd client before acquiring name.  Fixes logind race.
	* Add SessionName property to introspection. Fixes gdbus errors
	* drop system inhibitor when necessary. Fixes logind leak.
	* fix malloc/g_free mismatch in systemd code
	* Updated translations

=============
Version 3.6.1
=============

 Session Manager

	* Don't fall back for zaphod mode multi-screen setups
	* Fix build on OpenBSD
	* Detect greeter sessions better with systemd
	* Updated translations

=============
Version 3.6.0
=============

 Session Manager

	* Make autostart overriding more flexible
	* Updated translations

==============
Version 3.5.91
==============

 Session Manager

	* Updated translations

=============
Version 3.5.5
=============

 Session Manager

	* Make XSMP error messages more clear
	* Be more careful in systemd code when session fails to get registered
	* Require newer glib
	* Make gconf build time optional
	* Updated translations

=============
Version 3.5.4
=============

 Session Manager

	* Add command for showing reboot dialog
	* Inhibitor API / mechanism improvements
        * Systemd integration fixes
	* man page updates
	* fail whale in more failure cases
	* fix duplicate logout dialog when user hits cancel
	* Updated translations

=============
Version 3.5.2
=============

 Session Manager

	* Require new logind version (Colin Walters)
	* Remove markup from translatable string (Piotr Drąg)
	* Fail softer in the face of spotty session bus (Ray, William Jon McCann)
	* Consider that required component that exits with 1 has crashed (Vincent)

  Translators

	* Dirgita (id)

=============
Version 3.4.2
=============

  Capplet

	* Add Unity to OnlyShowIn in .desktop file (Jeremy Bicha)

  Translators

	* Yurek Hinz (csb)
	* Takayuki KUSANO (ja)

=============
Version 3.4.1
=============

  Session Manager

	* Properly move to next phase if an app dies on startup (Vincent)
	* Fix some restart/poweroff problems in the systemd code (Matthias
	  Clasen)
	* Cleanups (Vincent)

  Translators

	* David Planella (ca)
	* David Planella (ca@valencia)
	* Reşat SABIQ (crh)
	* Wolfgang Stoeggl (de)
	* Bruce Cowan (en_GB)
	* Arash Mousavi (fa)
	* Gabor Kelemen (hu)
	* Milo Casagrande (it)
	* Sandeep Shedmake (mr)
	* Kjartan Maraas (nb)

=============
Version 3.4.0
=============

  Session Manager

	* Don't show toggle switches for gnome-shell extensions in fail whale
	  dialog; simply disable all extensions instead (Jasper St. Pierre)
	* Code cleanups (Jasper St. Pierre)

  Translators

	* Nilamdyuti Goswami (as)
	* Ihar Hrachyshka (be)
	* Alexander Shopov (bg)
	* David Planella (ca)
	* David Planella (ca@valencia)
	* Reşat SABIQ (crh)
	* Marek Černocký (cs)
	* flemming christensen (da)
	* Daniel Mustieles (es)
	* Mattias Põldaru (et)
	* Luc Guillemin (fr)
	* Fran Dieguez (gl)
	* Sweta Kothari (gu)
	* Yaron Shahrabani (he)
	* Changwoo Ryu (ko)
	* Aurimas Černius (lt)
	* Rūdolfs Mazurs (lv)
	* A S Alam (pa)
	* Piotr Drąg (pl)
	* Duarte Loreto (pt)
	* Antonio Fernandes C. Neto (pt_BR)
	* Yuri Myasoedov (ru)
	* Matej Urbančič (sl)
	* Мирослав Николић (sr)
	* Miroslav Nikolić (sr@latin)
	* Daniel Nylander (sv)
	* Sasi Bhushan Boddepalli (te)
	* Korostil Daniel (uk)
	* Nguyễn Thái Ngọc Duy (vi)
	* tuhaihe (zh_CN)
	* Chao-Hsiung Liao (zh_HK)
	* Chao-Hsiung Liao (zh_TW)

==============
Version 3.3.92
==============

  Session Manager

	* Minor bug fixes (Ray Strode, Jasper St. Pierre)

  Translators

	* Nilamdyuti Goswami (as)
	* Mattias Põldaru (et)
	* Yuri Myasoedov (ru)

==============
Version 3.3.90
==============

  Session Manager

	* Fix the GsmSystem type registration (Matthias Clasen)

=============
Version 3.3.5
=============

  Session Manager

	* Use systemd session tracking instead of ConsoleKit when system is
	  booted with systemd (Matthias Clasen)
	* Code cleanups (Matthias Clasen)

  Translators

	* Hideki Yamane (ja)
	* Andreas N. (nn)
	* Muhammet Kara (tr)

=============
Version 3.3.3
=============

  Tools

	* Blacklist pre-R300 Radeon hardware in accelerated check (Vincent)

=============
Version 3.3.2
=============

  Tools

	* Fix regexp for blacklisting Intel 830-865 hardware (Vincent)

  Translators

	* Kristjan SCHMIDT (eo)
	* Muhammet Kara (tr)

=============
Version 3.3.1
=============

  Tools

	* Look at gnome.fallback={0,1} argument in kernel boot line to
	  determine if fallback mode should be enforced or ignored (Vincent)
	* Use whitelist/blacklist regexps in an external file when evaluating
	  renderer for accelerated check (Vincent)
	* Blacklist Intel 830-865 hardware when checking for accelerated
	  hardware (Vincent)

  Translators

	* Ihar Hrachyshka (be)

=============
Version 3.2.1
=============

  Session Manager

	* Refuse to install an idle timeout of 0 seconds (Rodrigo Moya)
	* Fix crash on Solaris because of bad comparison (Brian Cameron)

  Translators

	* David Planella (ca@valencia)
	* Kristjan SCHMIDT (eo)
	* Nguyễn Thái Ngọc Duy (vi)
	* Chao-Hsiung Liao (zh_HK)
	* Chao-Hsiung Liao (zh_TW)

=============
Version 3.2.0
=============

  Translators

	* Nilamdyuti Goswami (as)
	* Petr Kovar (cs)
	* flemming christensen (da)
	* Inaki Larranaga Murgoitio (eu)
	* Tommi Vainikainen (fi)
	* Sweta Kothari (gu)
	* Gabor Kelemen (hu)
	* Andika Triwidada (id)
	* Changwoo Ryu (ko)
	* Aurimas Černius (lt)
	* Мирослав Николић (sr)
	* Miroslav Nikolić (sr@latin)
	* Praveen Illa (te)
	* Korostil Daniel (uk)

==============
Version 3.1.92
==============

  Session Manager

	* Never mark as handled the Disconnected signal from dbus (Vincent)
	* Disconnect all dbus clients when dbus is disconnected (Vincent)
	* Don't offer logout in the fail whale on the login screen (Matthias Clasen)
	* Code improvements (Matthias Clasen, Vincent)

  Translators

	* Anass Ahmed (ar)
	* Ihar Hrachyshka (be)
	* Alexander Shopov (bg)
	* David Planella (ca)
	* Mario Blättermann (de)
	* Bruce Cowan (en_GB)
	* Daniel Mustieles (es)

tools/gnome-session-quit.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*-
 * save-session.c - Small program to talk to session manager.

   Copyright (C) 1998 Tom Tromey
   Copyright (C) 2008 Red Hat, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
   02111-1307, USA.
*/

#include <config.h>

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <glib/gi18n.h>
#include <gtk/gtk.h>

#include <dbus/dbus-glib.h>
#include <dbus/dbus-glib-lowlevel.h>

#define GSM_SERVICE_DBUS   "org.gnome.SessionManager"
#define GSM_PATH_DBUS      "/org/gnome/SessionManager"
#define GSM_INTERFACE_DBUS "org.gnome.SessionManager"

enum {
        GSM_LOGOUT_MODE_NORMAL = 0,
        GSM_LOGOUT_MODE_NO_CONFIRMATION,
        GSM_LOGOUT_MODE_FORCE
};

static gboolean opt_logout = FALSE;
static gboolean opt_power_off = FALSE;
static gboolean opt_reboot = FALSE;
static gboolean opt_no_prompt = FALSE;
static gboolean opt_force = FALSE;

static GOptionEntry options[] = {
        {"logout", '\0', 0, G_OPTION_ARG_NONE, &opt_logout, N_("Log out"), NULL},
        {"power-off", '\0', 0, G_OPTION_ARG_NONE, &opt_power_off, N_("Power off"), NULL},
        {"reboot", '\0', 0, G_OPTION_ARG_NONE, &opt_reboot, N_("Reboot"), NULL},
        {"force", '\0', 0, G_OPTION_ARG_NONE, &opt_force, N_("Ignoring any existing inhibitors"), NULL},
        {"no-prompt", '\0', 0, G_OPTION_ARG_NONE, &opt_no_prompt, N_("Don't prompt for user confirmation"), NULL},
        {NULL}
};

static void
display_error (const char *message)
{
        g_printerr ("%s\n", message);
}

static DBusGConnection *
get_session_bus (void)
{
        DBusGConnection *bus;
        GError *error = NULL;

        bus = dbus_g_bus_get (DBUS_BUS_SESSION, &error);

        if (bus == NULL) {
                g_warning ("Couldn't connect to session bus: %s", error->message);
                g_error_free (error);
        }

        return bus;
}

static DBusGProxy *
get_sm_proxy (void)
{
        DBusGConnection *connection;
        DBusGProxy      *sm_proxy;

        connection = get_session_bus ();
        if (connection == NULL) {
                display_error (_("Could not connect to the session manager"));
                return NULL;
        }

        sm_proxy = dbus_g_proxy_new_for_name (connection,
                                               GSM_SERVICE_DBUS,
                                               GSM_PATH_DBUS,
                                               GSM_INTERFACE_DBUS);

        if (sm_proxy == NULL) {
                display_error (_("Could not connect to the session manager"));
                return NULL;
        }

        return sm_proxy;
}

static void
do_logout (unsigned int mode)
{
        DBusGProxy *sm_proxy;
        GError     *error;
        gboolean    res;

        sm_proxy = get_sm_proxy ();
        if (sm_proxy == NULL) {
                return;
        }

        error = NULL;
        res = dbus_g_proxy_call (sm_proxy,
                                 "Logout",
                                 &error,
                                 G_TYPE_UINT, mode,
                                 G_TYPE_INVALID,
                                 G_TYPE_INVALID);

        if (!res) {
                if (error != NULL) {
                        g_warning ("Failed to call logout: %s",
                                   error->message);
                        g_error_free (error);
                } else {
                        g_warning ("Failed to call logout");
                }
        }

        if (sm_proxy != NULL) {
                g_object_unref (sm_proxy);
        }
}

static void
do_power_off (const char *action)
{
        DBusGProxy *sm_proxy;
        GError     *error;
        gboolean    res;

        sm_proxy = get_sm_proxy ();
        if (sm_proxy == NULL) {
                return;
        }

        error = NULL;
        res = dbus_g_proxy_call (sm_proxy,
                                 action,
                                 &error,
                                 G_TYPE_INVALID,
                                 G_TYPE_INVALID);

        if (!res) {
                if (error != NULL) {
                        g_warning ("Failed to call %s: %s",
                                   action, error->message);
                        g_error_free (error);
                } else {
                        g_warning ("Failed to call %s", action);
                }
        }

        if (sm_proxy != NULL) {
                g_object_unref (sm_proxy);
        }
}

int
main (int argc, char *argv[])
{
        GError *error;
        int     conflicting_options;

        /* Initialize the i18n stuff */
        bindtextdomain (GETTEXT_PACKAGE, LOCALE_DIR);
        bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
        textdomain (GETTEXT_PACKAGE);

        error = NULL;
        if (! gtk_init_with_args (&argc, &argv, NULL, options, NULL, &error)) {
                g_warning ("Unable to start: %s", error->message);
                g_error_free (error);
                exit (1);
        }

        conflicting_options = 0;
        if (opt_logout)
                conflicting_options++;
        if (opt_power_off)
                conflicting_options++;
        if (opt_reboot)
                conflicting_options++;
        if (conflicting_options > 1)
                display_error (_("Program called with conflicting options"));

        if (opt_power_off) {
                do_power_off ("Shutdown");
        } else if (opt_reboot) {
                do_power_off ("Reboot");
        } else {
                /* default to logout */

                if (opt_force)
                        do_logout (GSM_LOGOUT_MODE_FORCE);
                else if (opt_no_prompt)
                        do_logout (GSM_LOGOUT_MODE_NO_CONFIRMATION);
                else
                        do_logout (GSM_LOGOUT_MODE_NORMAL);
        }

        return 0;
}
